In [8]:
import pandas as pd

from sklearn.decomposition import IncrementalPCA
from scipy.stats.stats import pearsonr

1) Загрузите данные close_prices.csv. В этом файле приведены цены акций 30 компаний на закрытии торгов за каждый день периода. 

In [6]:
data_train = pd.read_csv(r'/Users/u17491440/PycharmProjects/enterInML/data/week_4/close_prices.csv')
print(data_test.head())

data_test = pd.read_csv(r'/Users/u17491440/PycharmProjects/enterInML/data/week_4/djia_index.csv')
print(data_train.head())

         date        AXP          BA        CAT       CSCO         CVX  \
0  2013-09-23  76.440002  117.510002  85.029999  24.270000  125.519997   
1  2013-09-24  76.070000  119.000000  85.110001  24.139999  124.489998   
2  2013-09-25  75.989998  118.510002  84.500000  24.430000  124.070000   
3  2013-09-26  76.320000  119.379997  84.199997  23.770000  123.489998   
4  2013-09-27  75.889999  118.739998  83.800003  23.330000  122.639999   

          DD        DIS         GE          GS    ...            PFE  \
0  59.409999  64.750000  24.280001  165.250000    ...      28.799999   
1  59.319997  64.320000  24.320000  162.970001    ...      28.709999   
2  59.319997  64.449997  24.230000  162.309998    ...      28.490000   
3  59.509996  65.239998  24.250000  162.289993    ...      28.520000   
4  59.009995  65.190002  24.049999  159.850006    ...      28.879999   

          PG          T        TRV        UNH         UTX           V  \
0  79.279999  34.220001  86.379997  71.820000  10

2) На загруженных данных обучите преобразование PCA с числом компоненты равным 10. Скольких компонент хватит, чтобы объяснить 90% дисперсии?

In [7]:
ipca = IncrementalPCA(n_components=10)
# Обучающая выборка
del data_train['date']
ipca.fit(data_train)
print(ipca.explained_variance_ratio_)

[0.73892531 0.11003496 0.04989187 0.02854979 0.02211184 0.019276
 0.0066939  0.0059554  0.00306374 0.00297776]


3) Примените построенное преобразование к исходным данным и возьмите значения первой компоненты. 

4) Загрузите информацию об индексе Доу-Джонса из файла djia_index.csv. Чему равна корреляция Пирсона между первой компонентой и индексом Доу-Джонса? 

In [12]:
pearsonr(ipca.transform(data_train)[:,0], data_test['^DJI'])

(0.9096590977185374, 5.123261330854433e-144)

5) Какая компания имеет наибольший вес в первой компоненте? Укажите ее название с большой буквы.

In [13]:
max_corr = 0
company = 0
for i in range(0,30,1):
    if (pearsonr(ipca.transform(data_train)[:,0], data_train.iloc[:,i])[0] >  max_corr):
        max_corr = pearsonr(ipca.transform(data_train)[:,0], data_train.iloc[:,i])[0]
        company = i
print(list(data_train.columns.values)[company]) 

V


Сколько компонент достаточно, чтобы описать 90% дисперсии?
3
Вычислите коэффициент корреляции Пирсона между значением по первой компоненте и индексом Доу-Джонса.
0.91
Какая компания имеет наибольший вес в первой компоненте?
V